File: speed_tests/linalg_tests/python/mat.py
import numpy as np
from numpy import linalg as LA
from common import *


@timer(cnt=count)
def create_random_matrix(shape: tuple) -> np.ndarray:
    """
    Создает матрицу с рандомными элементами [0;100)
    :param n: int - shape.
    :return:
    - <class 'numpy.ndarray'> - рандомная матрица.
    """

    return np.random.rand(shape[0]*shape[1]).reshape(shape[0], shape[1]) * 100


@timer(cnt=count)
def scale_matrix(a: np.ndarray, alpha: float):
    """
    Масштабирует переданную матрицу на alpha.
    :param a: np.ndarray - матрица для масштабирования;
    :param alpha: float - параметр масштабирования.
    """

    a * alpha


@timer(cnt=count)
def transpose_matrix(a: np.ndarray) -> np.ndarray:
    """
    Транспонирует переданную матрицу.
    :param a: np.ndarray - матрица для транспонирования.
    :return:
    -     - <class 'numpy.ndarray'> - транспонированная матрица a.
    """

    return a.transpose()


@timer(cnt=count)
def addition_of_matrices(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    """
    Считает сумму двух матриц.
    :param a: np.ndarray - слагаемое;
    :param b: np.ndarray - слагаемое.
    :return:
    - np.ndarray - сумма.
    """

    return np.add(a, b)


@timer(cnt=count)
def subtract_of_matrices(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    """
    Считает разницу двух векторов.
    :param a: np.ndarray - уменьшаемое;
    :param b: np.ndarray - вычитаемое.
    :return:
    - np.ndarray - разность.
    """

    return np.subtract(a, b)


@timer(cnt=count)
def dot_of_matrices(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    """
    Считает произведение двух векторов.
    :param a: np.ndarray - множитель;
    :param b: np.ndarray - множитель.
    :return:
    - np.ndarray - произведение.
    """

    return np.dot(a, b)


@timer(cnt=count)
def determinant_of_matrix(a: np.ndarray) -> float:
    """
    Считает детерминант для переданной матрицы.
    :param a: np.ndarray - матрицы для поиска детерминанта.
    :return:
    - float - детерминант матрицы.
    """

    return LA.det(a)


@timer(cnt=count)
def eigens_of_matrix(a: np.ndarray) -> (np.ndarray, np.ndarray):
    """
    Находит собственные числа и правые собственные вектора.
    :param a: np.ndarray - матрицы для поиска собственных знач. и век..
    :return:
    - (np.ndarray, np.ndarray) - вектор собственных значений и
    матрица собственных векторов.
    """

    return LA.eig(a)


@timer(cnt=count)
def SVD_of_matrix(a: np.ndarray) -> (np.ndarray, np.diag, np.ndarray):
    """
    Находит элементы SVD разложения матрицы a.
    :param a: np.ndarray - матрицы для нахождения SVD разложения.
    :return:
    - (np.ndarray, np.diag, np.ndarray) - U, S, V'.
    """

    return LA.svd(a)


@timer(cnt=count)
def cholesky_of_matrix(a: np.ndarray) -> np.ndarray:
    """
    Находит разложение Холецкого для переданной эрмитовой
    (симметричной) положительно-определенной матрицы.
    :param a: np.ndarray - эрмитовая (симметричная)
    положительно-определенная матрица.
    :return:
    - np.ndarray - нижнетреугольная матрица L.
    """

    return LA.cholesky(a)

File: speed_tests/linalg_tests/go/mat.go
package speedtests

import (
	"fmt"
	"gonum.org/v1/gonum/mat"
	"log"
	"math/rand"
	"time"
)

// createRandomMatrix create NxM random Matrix [0;100)
func createRandomMatrix(N, M int) (A *mat.Dense) {
	v := make([]float64, N*M)
	for i := 0; i < N*M; i++ {
		rand.Seed(int64(time.Now().UnixNano()))
		v[i] = rand.Float64() * 100
	}
	A = mat.NewDense(N, M, v)
	return
}

// scaleMatrix make inplace scaling of given matrix
func scaleMatrix(alpha float64, A *mat.Dense) {
	A.Scale(alpha, A)
}

// transposingMatrix return interface Matrix with structure of
// transposed matrix
func transposeMatrix(A *mat.Dense) mat.Matrix {
	return A.T()
}

// additionOfMatrices return result of addition of two matrices
func additionOfMatrices(A, B *mat.Dense) (C *mat.Dense) {
	Rows, Cols := A.Dims()
	C = mat.NewDense(Rows, Cols, nil)
	C.Add(A, B)
	return
}

// subtractOfMatrices return result of subtract of two matrices
func subtractOfMatrices(A, B *mat.Dense) (C *mat.Dense) {
	Rows, Cols := A.Dims()
	C = mat.NewDense(Rows, Cols, nil)
	C.Sub(A, B)
	return
}

// dotOfMatrices return dot of matrices
func dotOfMatrices(A, B *mat.Dense) (C *mat.Dense) {
	Rows, Cols := A.Dims()
	C = mat.NewDense(Rows, Cols, nil)
	C.Product(A, B)
	return
}

// determinantOfMatrix return determinant of given matrix
func determinantOfMatrix(A *mat.Dense) (d float64) {
	d = mat.Det(A)
	return
}

// eigensOfMatrix returns right eigen vectors and values
func eigensOfMatrix(A *mat.Dense) (eigenValues []complex128, eigenVectors *mat.CDense){
	var eig mat.Eigen
	if ok := eig.Factorize(A, mat.EigenRight); !ok {
		log.Fatal("Eigendecomposition failed")
	}
	eigenValues = eig.Values(nil)
	eigenVectors = eig.VectorsTo(nil)
	return
}

// SVDOfMatrix make SVD decomposition of given matrix
func SVDOfMatrix(A *mat.Dense) (S []float64, U, V *mat.Dense) {
	var SVD mat.SVD
	if ok := SVD.Factorize(A, mat.SVDFull); !ok {
		log.Fatal("SVD failed")
	}
	S, U, V = extractSVD(&SVD)
	return
}

// choleskyOfMatrix returns L of A = L'L Cholesky decomposition
func choleskyOfMatrix(A *mat.SymDense) (L *mat.TriDense) {
	var cholesky mat.Cholesky
	if ok := cholesky.Factorize(A); !ok {
		log.Fatal("A matrix is not positive semi-definite.")
	}
	L = cholesky.LTo(nil)
	return
}

// matPrint print Matrix to Stdout
func matPrint(A mat.Matrix) {
	fmt.Printf("%v\n", mat.Formatted(A, mat.Prefix(" "), mat.Excerpt(3)))
}

// extractSVD extracts SVD decomposition results
func extractSVD(svd *mat.SVD) (s []float64, u, v *mat.Dense) {
	return svd.Values(nil), svd.UTo(nil), svd.VTo(nil)
}
